common/util/auto_pop_stack.h

```diff
@@ -0,0 +1,79 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_COMMON_UTIL_AUTO_POP_STACK_H_
+#define VERIBLE_COMMON_UTIL_AUTO_POP_STACK_H_
+
+#include <vector>
+
+namespace verible {
+
+// AutoPop stack class. Templated version of SyntaxTreeContext::AutoPop
+template <typename T>
+class AutoPopStack {
+ public:
+  typedef T value_type;
+  typedef AutoPopStack<value_type> this_type;
+
+  class AutoPop {
+   public:
+    AutoPop(this_type* stack, T& value) {
+      stack_ = stack;
+      stack->Push(value);
+    }
+    ~AutoPop() {
+      stack_->Pop();
+    }
+
+   private:
+    this_type* stack_;
+  };
+
+  bool empty() {
+    return stack_.empty();
+  }
+
+  value_type& top() {
+    CHECK(!stack_.empty());
+    return *ABSL_DIE_IF_NULL(stack_.back());
+  }
+
+  typedef std::vector<value_type*> stack_type;
```

Comment:
`typedef std::vector<value_type> stack_type;`

By removing the pointer from this implementation, you make this a suitable base class for `SyntaxTreeContext`.

```c++
class SyntaxTreeContext : public AutoPopStack<const SyntaxTreeNode*> {
  ...
 public:
  using AutoPop = base_type::AutoPop;
  ...
};
```

The only adjustment you might have to make is that in the SyntaxTreeContext subclass `top()` (or renamed) should return a `const T&` (or `const SyntaxTreeNode*&`) to be agnostic to whether T is a pointer type or object type.

From a mutability perspective, `SyntaxTreeContext` was designed to forbid any mutation to the stack elements, even the top element -- everything is const, and `AutoPop` is the *only* interface for mutation.  But I'm aware you intend to use it in this PR with a mutable top/current scope, because you have to register declared symbols.  So as far as class design is concerned, the common base class may support mutable elements, but the subclasses may choose to further restrict the interfaces for what is mutable (where private inheritance can help).  

Let's say your scope-stack subclass wants to ensure that only the top-most scope can be modified, but all others can be looked-up with const accesses.  You could lock down most of your interface by exposing only const iterators/methods (hiding the mutable ones), except for `T& top()` as the only mutable entry point.  The base class could provide both const and non-const top(), and the subclass could present only the const one if it chose to.

Keep this in mind should you continue to refactor this.

Reply:
I did write this class in mind that it will be used in `SyntaxTreeContext`. Thanks for noticing that. I think that it will be easier to work on this class in separate PR #289